tools/get-images.sh

```diff
@@ -38,6 +39,13 @@ NET_ISTIO_VERSION="${KNATIVE_SERVING_VERSION}"
 if [ $KNATIVE_SERVING_VERSION == "1.10.2" ]; then
   NET_ISTIO_VERSION="1.11.0"
 fi
+
+# For Serving 1.12.4 (CKF 1.9) we have to use 1.12.3 for net-istio
+# https://github.com/kubeflow/manifests/pull/2709
+if [ $KNATIVE_SERVING_VERSION == "1.12.4" ]; then
+  NET_ISTIO_VERSION="1.12.3"
+fi
+
```

Comment:
There are a couple of things happening, actually.

1. `NET_ISTIO_VERSION` is set to `KNATIVE_SERVING_VERSION` - I think that by design, this script assumed that there would be a 1:1 mapping of the versions of net-istio and knative-serving.
2. If the knative-serving version happened to be `1.10.2`, the net-version would've been set to a hard coded value. I guess it was done this way because the script didn't always work with the version at the time this script was created and this was added as a hack. I see [L36-40](https://github.com/canonical/knative-operators/blob/main/tools/get-images.sh#L36C1-L40C3) are based on what the kubeflow/manifests repo would have for versions at the moment of release (for example, for 1.8 [here](https://github.com/kubeflow/manifests/tree/v1.8-branch/common/knative#knative-serving)).
3. From the comment in L36 I can tell the hack was intended to be updated on every release if needed.

Based on the above, I think that:

1. We do not need to add a condition per knative-version that we have. These scripts are used by our CI to gather the images from a branch, so `track/1.8` would have a very different version of knative-version than `track/1.9`, for example, so we always know the value of KNATIVE_SERVING_VERSION.
2. Maybe there is a better way to fetch the version of net-istio automatically. From the [README](https://github.com/kubeflow/manifests/blob/v1.9-branch/common/knative/README.md?plain=1#L8)(?), though we can look for a better source of truth.
3. If (2) is not possible, this looks like a chore that we were not aware during our upgrades for 1.9. I think we have to make it more clear that this file would have to be updated as well on every release. My ask would be to add it to the CONTRIBUTING.md so our team is aware.

Reply:
Regarding the initial design decisions, I've linked that issue in the issue of the PR for extra context.

So based on that:
> 1. We do not need to add a condition per knative-version that we have. These scripts are used by our CI to gather the images from a branch, so track/1.8 would have a very different version of knative-version than track/1.9, for example, so we always know the value of KNATIVE_SERVING_VERSION

Yes, good point. I'll update to only have an if statement for the version for the Knative Serving version we expect in the branch

> 2. Maybe there is a better way to fetch the version of net-istio automatically. From the [README](https://github.com/kubeflow/manifests/blob/v1.9-branch/common/knative/README.md?plain=1#L8)(?), though we can look for a better source of truth

You raise a good point, that the problem with `net-istio` compared to Serving, is that for Serving and Eventing we explicitly [define the version](https://github.com/canonical/knative-operators/blob/main/charms/knative-serving/config.yaml#L6) which is what this script uses.

But for `net-istio` we don't have this in the charms, and end up hardcoding it. So for me the question then is, where in our charms should we expose this version so then the script know where to look for it? (and in the future the scrip will gather the images from the charms, rather than fetching the manifests https://github.com/canonical/knative-operators/issues/220)

> 3. If (2) is not possible, this looks like a chore that we were not aware during our upgrades for 1.9. I think we have to make it more clear that this file would have to be updated as well on every release. My ask would be to add it to the CONTRIBUTING.md so our team is aware.

If we move and have this version information to the Charm, then we won't have to make any extra chore aside from updating the versions which we either way would have to do.

Reply:
> Regarding my points on `net-istio` version being deduced, this should be resolved on #220

Agreed. Just so it is registered here as well, right now these are deduced by checking in the upstream documentation. This is definitely not scalable and prone to error, so it's a good thing you have #220 for this.

I think then you have taken a good approach for now until we have a better way of getting these images.